backends/ebpf/runtime/ebpf_runtime.c

```diff
@@ -14,89 +14,270 @@ See the License for the specific language governing permissions and
+        }
+        /* Reset the file pointer and reopen the pcap handle */
+        fseek(pcap_in_fp, 0, 0);
+        in_handles[i] = pcap_fopen_offline(pcap_in_fp, errbuf);
+        if (in_handles[i] == NULL) {
+            fprintf(stderr, "Error: Failed to reopen pcap file!\n");
+            exit(EXIT_FAILURE);
+        }
+        /* Allocate a list of packets and reset the index */
+        pcap_pkt *pkts = malloc(sizeof(pcap_pkt) * pkt_index);
+        pkt_index = 0;
+        /* Fill the packet list while incrementing the counter */
+        while ((ret = pcap_next_ex(in_handles[i], &pcap_hdr, &packet)) == 1) {
+            /* Save the data we extracted from the pcap buffer */
+            pkts[pkt_index].data = malloc(pcap_hdr->len);
+            memcpy(pkts[pkt_index].data, packet, pcap_hdr->len);
+            /* Also save the header and the interface "index" */
+            pkts[pkt_index].pcap_hdr = *pcap_hdr;
+            pkts[pkt_index].ifindex = i;
+            pkt_index++;
+        }
+        if (ret == -1) {
+            pcap_perror(in_handles[i], "Error: Failed to parse ");
+            exit(EXIT_FAILURE);
+        }
+        /* Store the pointer to the list in the packet container */
+        pkt_list[i].list = pkts;
+        pkt_list[i].len = pkt_index;
+    }
+}
+
+static void free_pkt_list(pcap_pkt_list *pkt_list) {
+    for (uint64_t i = 0; i < pkt_list->len; i++) {
+        free(pkt_list->list[i].data);
+    }
+    free(pkt_list->list);
+    free(pkt_list);
+}
+
+static pcap_pkt_list *get_sorted_pkts(pcap_t **in_handles, int num_files) {
+    pcap_pkt_list pcap_pkt_lists[num_files];
+    /* Fill the packet lists with packets from the pcap files */
+    fill_packets(pcap_pkt_lists, num_files, in_handles);
+    /* Merge, sort, and return the list of the parsed packets */
+    return sort_packets(pcap_pkt_lists, num_files);
+}
+
+static void feed_packets(pcap_dumper_t **out_handles, pcap_pkt_list *pkt_list, int num_pcaps) {
+    for (uint64_t i = 0; i < pkt_list->len; i++) {
+        struct pcap_pkthdr *pcap_hdr = &pkt_list->list[i].pcap_hdr;
+        pcap_dumper_t *out_handle = out_handles[pkt_list->list[i].ifindex];
         /* Parse each packet in the file and check the result */
         struct sk_buff skb;
-        skb.data = (void *) packet;
-        skb.len = pcap_hdr->len;
+        skb.data = (void *) pkt_list->list[i].data;
+        skb.len = pkt_list->list[i].pcap_hdr.len;
         int result = ebpf_filter(&skb);
         if (result)
-            pcap_dump((unsigned char *) out_handle, pcap_hdr, packet);
+            pcap_dump((unsigned char *)out_handle, pcap_hdr, skb.data);
         if (debug)
             printf("\nResult of the eBPF parsing is: %d\n", result);
     }
-    return ret;
 }
 
-int main(int argc, char **argv) {
-    pcap_t *in_handle;
-    pcap_dumper_t *out_handle;
+static void init_output_handles(char *first, char *second, pcap_t **in_handles, pcap_dumper_t **out_handles, int num_files) {
```

Comment:
This function will disappear: it will be replaced by a for loop that calls the function to write a list to a file.
Parts of this code will move into the write function.